### ISSUE TITLE
Ability to override a target

### ISSUE THREAD

**Issue by wrouesnel** [restic/restic#555]
When doing backups with a filesystem which supports snapshots, the backup target and the filesystem path can change (often in uncontrollable ways).

i.e. if I'm backing up my home:
`/home/will`
but doing it with ZFS snapshots then I'm actually backing up:
`/home/will/.zfs/snapshot/restic-backup-1`

and that's what gets stored as the target. On a subsequent backup with a newly numbered snapshot, restic will then insist on re-reading the content of the entire filesystem i.e. if the snapshot mount is now `/home/will/.zfs/snapshot/restic-backup-2`

In both cases, what I really want to do is to specify the *intended* target and separate that concept from the mount - i.e. `/home/will/.zfs/snapshot/restic-backup-3` is actually just targeting `/home/will`

**Comment by fd0** [comment]
Hey, thanks for submitting this issue. It is in indeed an interesting feature to have.

I can imagine having sort of a "translation" rule, similar to `tar --transform` ([example](http://superuser.com/a/595512)), or just being able to strip path components and insert a custom prefix.

For your first example this could mean running restic as follows:
```
restic backup --prefix /home/will --strip-components 5 \
  /home/will/.zfs/snapshot/restic-backup-1
```
so that `/home/will/.zfs/snapshot/restic-backup-1/foo.txt` has the leading 5 path elements removed to `foo.txt` and then the prefix is applied, so that restic saves `/home/will/foo.txt`.

Is that what you meant?

This feature depends on the solution of #549, because there we'll probably change the way restic handles paths.

**Comment by wrouesnel** [comment]
That would pretty much do it I think.

**Comment by rumpelsepp** [comment]
I posted a similar request, maybe a duplicate here: https://github.com/restic/restic/issues/658

**Comment by martin21** [comment]
fd0, as you found the cause for issue #652 you hinted me at this issue and I thought it would be exactly what I want… yet… thats not entirely true, cause my BTRFS snapshots are in a *different* base path. For example:

1) /daten is the original directory
2) /mnt/daten-zeit/daten-resticbackup-$DATE or $UID would be unique snapshot directory name

Right now I work around issue #652 by providing a fixed snapshot name like "/mnt/daten-zeit/daten-resticbackup" without date. This appears to work well.

I prefer having all snapshots separately from the subvolume I snapshot.

So what I am after would be some "--pathinrepo" option or intended target option.

log "Sichere /home."
SNAP="/mnt/home-zeit/home-resticbackup-$DATE"
[ -d $SNAP ] && btrfs subvol del "$SNAP"
btrfs subvol snap /home "$SNAP"
restic backup \
        "$SNAP" \
        --pathinrepo "/home"
[…]
| stdbuf -oL -eL tee -a "$LOG"
log "/home gesichert."
btrfs subvol delete "$SNAP"

(note the --pathinrepo option in there)

**Comment by fd0** [comment]
Please read https://github.com/restic/restic/issues/555#issuecomment-236354707 again, assuming such a feature is implemented your use case is covered by:
```
$ restic backup --prefix /home --strip-components 3 \
   /mnt/home-zeit/home-resticbackup-$DATE
```

**Comment by martin21** [comment]
Ah, okay… I didn´t understand how  `--prefix` and `--strip-components` work together. This would work for me.

But I think its a bit error prone as the user need to keep `--strip-components 3` and snapshot path somewhat in sync. I.e. whenever I may change the snapshot path to have more or less levels I need to adapt the `--stip-components` option as well. But on the other hand that would be more flexible and explicit than "--pathinrepo" by splitting the operation into two parts: Remove something from the original path and then prepend a prefix. I´d just have gone with replacing the original path completely with `--pathinrepo` :)

**Comment by hardikar** [comment]
@fd0 You might also like to consider using the style that rsync does, which I feel is more concise : 
```
rsync -avR /foo/./bar/baz.c remote:/tmp/
That would create /tmp/bar/baz.c on the remote machine. (Note that the
    dot must be followed by a slash, so lq/foo/.rq would not be abbreviated.)
```
https://linux.die.net/man/1/rsync

Also noticed that I'm reviving an old issue, not sure if it has already been implemented in some way, I wasn't able to find `--prefix` or ` --strip-components` in the code.

**Comment by rawtaz** [comment]
I second that (again) - the rsync style of controlling the "target" part of the path is really really good. I used it many times with good success.

**Comment by mathiasnagler** [comment]
This would be very nice to have.

**Comment by lukastribus** [comment]
Indeed a very useful feature (also working with btrfs snapshots here), but this is useful for other use-cases as well: what if we want to upload from different places (maybe seed the backup from an external drive with a different path in a site with fast upload speeds). Currently, if the path doesn't match, restic has to rehash everything (admittedly though we don't need to reupload everything due do deduplication - still being able to specify the target would be *very* useful).

@fd0 now that the new archiver code is in, would it be the right time to take a look at this and ... is there room on your todo list for this :) ?

**Comment by giacomocariello** [comment]
Having some spare time, I did an initial attempt at this. I think there's still something to be corrected: If I try to subsequently create two snapshots of two distinct directories with the same content and fix the paths so that they appear to be the same directory, the "restic diff" command shows no difference but the report at the end of the backup shows as if all Files were created new.

**Comment by j6s** [comment]
I have looked into this a bit in the context of #2092.

My first attempt of implementing a flag which I called `--strip-prefix` in my tests where successful: In `archiver.Save` the `abstarget` variable can be modified in order to change the path that is saved into the repository without changing any other behaviour.

There is one problem though - and that is existing nodes in the repository:
If a node has not changed since the last backup then it is left in place - including its metadata (since those have also not been changed).

- Assuming you backup `/folder/with/prefix/data`
- Now this new feature comes around and allows you to strip that annoying prefix
- Every unchanged file will still be saved in the repository as `/folder/with/prefix/data/old-file.txt` whereas new and changed files will be in `/new-file.txt`. That's no good.

TL;DR It's not a easy as adjusting a single variable - we have to take existing nodes in repositories into account.

**Comment by ifedorenko** [comment]
> * Every unchanged file will still be saved in the repository as /folder/with/prefix/data/old-file.txt whereas new and changed files will be in /new-file.txt. That's no good.

Are you sure about it? I thought `Archiver.Snapshot` would not be able to match new nodes without the prefix with existing prefixed nodes. So the first unprefixed snapshot will have to re-read all files locally and will create new unprefixed nodes in the repository, but I think that's acceptable.

**Comment by j6s** [comment]
I am very sure about that, yes.

- I created a repository and made a single backup of `/home/j/tmp/test_source` using no flag
- I created a new backup using the flag to strip away the `/home/j/tmp/` part
- The snapshot created by the second backup contained
    - All unchanged files under `/home/j/tmp/test_source`
    - All changed files under `/test_source`

**Comment by dippynark** [comment]
A fairly easy workaround is to create a 'latest' snapshot -- a snippet from my backup script using BTRFS:
```
subvolumes=(/mnt/hdd/data /mnt/ssd/data)

# Snapshot BTRFS subvolumes
date=`date +"%FT%T"`
for subvolume in ${subvolumes[@]}; do
  # Create snapshot
  mkdir -p "$(dirname $subvolume)/snapshots/$(basename $subvolume)"
  btrfs subvolume snapshot $subvolume "$(dirname $subvolume)/snapshots/$(basename $subvolume)/$date"
  touch "$(dirname $subvolume)/snapshots/$(basename $subvolume)/$date"

  # Create latest snapshot
  ls "$(dirname $subvolume)/snapshots/$(basename $subvolume)/latest" && btrfs subvolume delete -c "$(dirname $subvolume)/snapshots/$(basename $subvolume)/latest"
  btrfs subvolume snapshot "$(dirname $subvolume)/snapshots/$(basename $subvolume)/$date" "$(dirname $subvolume)/snapshots/$(basename $subvolume)/latest"

  # Prune snapshots
  cd "$(dirname $subvolume)/snapshots/$(basename $subvolume)"
  ls -1rt "$(dirname $subvolume)/snapshots/$(basename $subvolume)" | grep -v latest | head -n -8 | xargs -r btrfs subvolume delete -c --
done
```

Then restic can just target the latest snapshot for backup.
